File: src/torch_mpc/cost_functions/generic_cost_function.cpp
#include "generic_cost_function.h"
#include "cost_terms/euclidean_distance_to_goal.h"
#include "cost_terms/footprint_costmap_projection.h"
#include "cost_terms/footprint_speedmap_projection.h"

int main() {

    // This script might need adjustment to work properly
    // See what inputs are required for the cost function classes
    std::vector<std::pair<double, std::shared_ptr<CostTerm>>> cost_terms = {
       {2.0, std::make_shared<EuclideanDistanceToGoal>()},
       // {1.0, std::make_shared<FootprintSpeedmapProjection>()},
       // {1.0, std::make_shared<FootprintCostmapProjection>()}
    };

    torch::Device device = torch::kCPU;  // try torch::kCUDA later

    CostFunction cfn(cost_terms, device);

    Values val;

    std::cout << "Check 0" << std::endl;
    std::cout << cfn.can_compute_cost() << std::endl;
    std::cout << "Expected: 0" << std::endl;

    // Simulate data loading
    torch::Tensor goal1 = torch::tensor({{5.0, 0.0}, {10.0, 0.0}});
    torch::Tensor goal2 = torch::tensor({{3.0, 0.0}, {4.0, 0.0}});
    torch::Tensor goal3 = torch::tensor({{6.0, 0.0}, {4.0, 0.0}});

    auto goals= torch::stack({goal1, goal2, goal3}, 0);
    val.data = goals;
    
    cfn.data.keys["waypoints"] = val;

    std::cout << "Check 1" << std::endl;
    std::cout << cfn.can_compute_cost() << std::endl;
    std::cout << "Expected: 0" << std::endl;

    Values val2 = val;
    cfn.data.keys["goals"] = val2;

    std::cout << "Check 2" << std::endl;
    std::cout << cfn.can_compute_cost() << std::endl;
    std::cout << "Expected: 0" << std::endl;
    
    std::unordered_map<std::string, torch::Tensor> metadata;
    metadata["resolution"] = torch::tensor({1.0, 0.5, 2.0});
    metadata["width"] = torch::tensor({100., 50., 200.});
    metadata["height"] = torch::tensor({100., 50., 200.});
    metadata["origin"] = torch::tensor({{-50., -50.}, {-25., -25.}, {-100., -100.}});
    metadata["length_x"] = torch::tensor({100., 50., 200.}); //might neeed adjustment
    metadata["length_y"] = torch::tensor({100., 50., 200.}); //might neeed adjustment

    Values val3 = val;
    val3.data = torch::zeros({1, 100, 100});
    val3.metadata = metadata;

    cfn.data.keys["local_costmap"] = val3;

    std::cout << "Check 3" << std::endl;
    std::cout << cfn.can_compute_cost() << std::endl;
    std::cout << "Expected: 0" << std::endl;

    Values val4 = val3;
    cfn.data.keys["local_speedmap"] = val3;

    std::cout << "Check 4" << std::endl;
    std::cout << cfn.can_compute_cost() << std::endl;
    std::cout << "Expected: 1" << std::endl;

    std::cout << "Check 5 (Final)" << std::endl;
    std::cout << "Plz don't crash" << std::endl;
    // Simulate states and actions
    auto states = torch::zeros({3, 4, 100, 5});
    states.index({torch::indexing::Slice(), 0, torch::indexing::Slice(), 0}) = torch::linspace(0, 60, 100);

    std::cout << "the states are created" << std::endl;

    auto actions = torch::zeros({3, 4, 100, 2});

    std::cout << "the actions are created" << std::endl;

    auto [costs, feasible] = cfn.cost(states, actions);
    std::cout << "Costs:\n" << costs << "\nFeasible:\n" << feasible << std::endl;

    return 0;
}

File: src/torch_mpc/cost_functions/generic_cost_function.py
import torch

class CostFunction:
    """
    High level cost-term aggregator that all MPC should be using to get the costs of trajectories

    In order to handle constraints properly, all cost terms must produce a feasible flag, in
        addition to a cost. We can then check if all trajs are infeasible and do something about it.

    Note that this feasible matrix is meant to be equivalent to allowing cost terms to prune out trajectories,
    but maintaining the batch shape.
    """
    def __init__(self, cost_terms, device='cpu'):
        """
        Args:
            cost_terms: a list of (weight, term) tuples that comprise the cost function
            data_timeout: time after which 
        """
        self.cost_terms = [x[1] for x in cost_terms]
        self.cost_weights = [x[0] for x in cost_terms]
        self.data = {k:None for k in self.get_data_keys()}

        self.device = device

    def cost(self, states, actions):
        """
        Produce costs from states, actions
        Args:
            states: a [B1 x B2 x T x N] tensor of states
            actions: a [B x B2 x T x M] tensor of actions
        """
        costs = torch.zeros(*states.shape[:2]).to(states.device)
        feasible = torch.ones(*states.shape[:2], dtype=bool).to(states.device)

        for cterm, cweight in zip(self.cost_terms, self.cost_weights):
            new_cost, new_feasible = cterm.cost(states, actions, feasible, self.data)

            # new_cost.to(self.device)
            # new_feasible.to(self.device)
            # costs.to(self.device)
            # feasible.to(self.device)
            
            costs += cweight * new_cost
            feasible = feasible & new_feasible

        return costs, feasible

    def get_data_keys(self):
        keys = set()
        for cterm in self.cost_terms:
            for label in cterm.get_data_keys():
                keys.add(label)

        return keys

    def can_compute_cost(self):
        return all([v is not None for v in self.data.values()])

    def to(self, device):
        self.device = device
        self.cost_terms = [x.to(device) for x in self.cost_terms]
        return self

    def __repr__(self):
        out = "Cost Function containing:\n"
        for cterm, cweight in zip(self.cost_terms, self.cost_weights):
            out += "\t{:.2f}:\t{}\n".format(cweight, cterm)
        return out

if __name__ == '__main__':
    cfn = CostFunction(
        cost_terms=[
            (2.0, EuclideanDistanceToGoal()),
            (1.0, CostmapProjection())
        ]
    )

    print(cfn.can_compute_cost())

    cfn.data['goals'] = [
        torch.tensor([
            [5.0, 0.0],
            [10.0, 0.0]
        ]),
        torch.tensor([
            [3.0, 0.0],
            [4.0, 0.0]
        ]),
        torch.tensor([
            [6.0, 0.0],
            [4.0, 0.0]
        ])
    ]

    print(cfn.can_compute_cost())

    cfn.data['costmap'] = torch.zeros(3, 100, 100)
    cfn.data['costmap'][:, 40:60, 60:] = 10.
    cfn.data['costmap_metadata'] = {
        'resolution':torch.tensor([1.0, 0.5, 2.0]),
        'width': torch.tensor([100., 50., 200.]),
        'height': torch.tensor([100., 50., 200.]),
        'origin': torch.tensor([[-50., -50.], [-25., -25.], [-100., -100.]])
    }

    states = torch.zeros(3, 4, 100, 5)
    states[:, 0, :, 0] = torch.linspace(0., 60., 100)
    actions = torch.zeros(3, 4, 100, 2)

    print(cfn.cost(states, actions))
